fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1338 @@
+ * limitations under the License.
+ */
+
+// This file implements the functions for RestoreMaster role
+
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/SystemData.h"
+#include "fdbclient/BackupAgent.actor.h"
+#include "fdbclient/ManagementAPI.actor.h"
+#include "fdbclient/MutationList.h"
+#include "fdbclient/BackupContainer.h"
+#include "fdbserver/RestoreCommon.actor.h"
+#include "fdbserver/RestoreRoleCommon.actor.h"
+#include "fdbserver/RestoreMaster.actor.h"
+#include "fdbserver/RestoreApplier.actor.h"
+#include "fdbserver/RestoreLoader.actor.h"
+
+#include "flow/actorcompiler.h"  // This must be the last #include.
+
+ACTOR Future<Void> askLoadersToCollectRestoreAppliersInterfaces(Reference<RestoreMasterData> self);
+ACTOR Future<Standalone<VectorRef<RestoreRequest>>> collectRestoreRequests(Database cx);
+ACTOR static Future<Version> processRestoreRequest(RestoreRequest request, Reference<RestoreMasterData> self, Database cx);
+ACTOR static Future<Void> finishRestore(Reference<RestoreMasterData> self, Database cx, Standalone<VectorRef<RestoreRequest>> restoreRequests);
+
+ACTOR static Future<Void> _collectBackupFiles(Reference<RestoreMasterData> self, Database cx, RestoreRequest request);
+ACTOR Future<Void> initializeVersionBatch(Reference<RestoreMasterData> self);
+ACTOR static Future<Void> distributeWorkloadPerVersionBatch(Reference<RestoreMasterData> self, Database cx, RestoreRequest request, Reference<RestoreConfig> restoreConfig);
+ACTOR static Future<Void> unlockDB(Database cx, UID uid);
+ACTOR static Future<Void> _clearDB(Reference<ReadYourWritesTransaction> tr);
+ACTOR static Future<Void> _lockDB(Database cx, UID uid, bool lockDB);
+ACTOR static Future<Void> registerStatus(Database cx, struct FastRestoreStatus status);
+ACTOR static Future<Void> sampleWorkload(Reference<RestoreMasterData> self, RestoreRequest request, Reference<RestoreConfig> restoreConfig, int64_t sampleMB_input);
+ACTOR Future<Void> notifyAppliersKeyRangeToLoader(Reference<RestoreMasterData> self, Database cx);
+ACTOR Future<Void> assignKeyRangeToAppliers(Reference<RestoreMasterData> self, Database cx);
+ACTOR Future<Void> notifyApplierToApplyMutations(Reference<RestoreMasterData> self);
+
+// The server of the restore master. It drives the restore progress with the following steps:
+// 1) Collect interfaces of all RestoreLoader and RestoreApplier roles
+// 2) Notify each loader to collect interfaces of all RestoreApplier roles
+// 3) Wait on each RestoreRequest, which is sent by RestoreAgent operated by DBA
+// 4) Process each restore request in actor processRestoreRequest;
+// 5) After process all restore requests, finish restore by cleaning up the restore related system key
+//    and ask all restore roles to quit.
+ACTOR Future<Void> startRestoreMaster(Reference<RestoreMasterData> self, Database cx) {
+	try {
+		wait( delay(1.0) );
+		wait( _collectRestoreRoleInterfaces(self, cx) );
```

Comment:
The actor that calls this one just did the recruitment, and knows what they are.  You could just pass it in to this actor instead of reading it from the database?